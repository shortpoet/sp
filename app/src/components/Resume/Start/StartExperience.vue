<template>
  <section class="resume-section p-3 px-lg-5 d-flex flex-column" id="experience">
    <div class="my-auto">
      <h2 class="heading heading-2 start-experience-header mb-4 mb-md-3">Experience</h2>
      <div class="resume-item d-flex flex-column mb-2" v-for="type in experiences" :key="type.type">
        <h2 class="heading heading-2 start-experience-type mb-3">{{ type.type }}</h2>
        <div class="resume-item d-flex flex-column flex-md-row mb-3" v-for="(job, ji) in type.jobs" :key="ji">
          <div class="resume-content mr-auto">
            <h3 class="heading heading-3 mb-0">{{ job.position }}</h3>
            <div class="start-company subheading mb-3">{{ job.company }}</div>
              <div
                v-for="(text, di) in job.description.split('\\n\\r')"
                :key="di"
              >
              <p v-html="text" :class="'job-' + (ji + 1) + '-description-' + (di + 1)"></p>
              </div>
          </div>
          <div class="resume-date text-md-right mt-1">
            <span v-if="hasDate(job.startDate)" class="text-primary">{{ job.startDate }} - {{ job.endDate }}</span>
          </div>
        </div>
      </div>
    </div>
  </section>
</template>
<script>
export default {
    name: 'StartExperience',
    props: {
      experiences: {
        type: Array
      }
    },
    methods: {
      hasDate: (date) => date !== '' && date !== null && date !== undefined
    }
}
</script>
