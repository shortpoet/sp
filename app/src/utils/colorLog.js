const chalk = require('chalk')
import {inspect} from 'util'

// const _log  = () => Function.prototype.bind.call(console.log, console, 'LOG_ALIAS')

export const log = (color, message) => console.log(chalk[`${color}`](`${inspect(message)}`))

export const colorLog = (message, color, background) => {
  color = color || "black";
  if (background) {
    console.log(`%c` + `${message}`, `color:` + `${color};background:${background}`)
  } else {
    console.log(`%c` + `${message}`, `color:` + `${color}`)
  }
}

// module.exports =  {
//   log,
//   colorLog
// }
