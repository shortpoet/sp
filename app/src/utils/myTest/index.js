import createWrapper from './test.utils.createWrapper'
import factory from './test.utils.factory'
import factoryShallow from './test.utils.factoryShallow'
import propFinder from './test.utils.propFinder'
import propsFinder from './test.utils.propsFinder'
import propMocker from './test.utils.propMocker'
import propsMocker from './test.utils.propsMocker'
import textMatcher from './test.utils.textMatcher'
import textMatcherFactory from './test.utils.textMatcherFactory'

export default {
  createWrapper,
  factory,
  factoryShallow,
  propFinder,
  propsFinder,
  propMocker,
  propsMocker,
  textMatcher,
  textMatcherFactory
}
