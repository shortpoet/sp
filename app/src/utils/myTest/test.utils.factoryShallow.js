import { shallowMount } from '@vue/test-utils'

export const factoryShallow = (
  component, 
  /* istanbul ignore next */
  options = {
    propsData: {},
    localVue: {},
    mocks: {},
    store: {},
    // context: {}, // only for functional components
    router: {},
    computed: {},
    stubs: {},
    slots: {},
    scopedSlots: {},
    attrs: {},
    attachToDocument: false,
    attachTo: null, //HTML Element or String
    listeners: {},
    parentComponent: {},
    provide: {}
  }) => {
    return shallowMount(component, options)
  }

module.exports = factoryShallow