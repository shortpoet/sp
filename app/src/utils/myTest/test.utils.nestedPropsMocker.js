

// // this is incomplete
// // maybe try and develop using TDD  
  
// export const nestedPropsMocker = (props) => {

//   console.log(props)

//   let propsData = {}


//   const handleArray = (props) => {
//     propsData = {}
//     props.map(prop => {
//       propsData[`${prop}`] = `sample ${prop}`
//     })
//     return {
//       propsData
//     }
//   }


//   const handleNested = (props, propsData, previousKey, nextKey) => {
//     console.log(props, propsData)

//     let firstLevel
//     if (previousKey === null && nextKey === null) {
//       firstLevel = true
//     }

//     // spread to clone so the splice works on subsequent tests
//     let values = [...props]

//     const length = parseInt(values.splice(-1, 1)[0])
//     console.log(length)
//     const iterator = [...Array(length).keys()]

//     if (firstLevel === true) {
//       previousKey = values.splice(0, 1)[0]
//       propsData[`${previousKey}`] = []
//     } else {
//       console.log('#### not first run ####')
//       console.log(propsData)
//       console.log(previousKey)
//       nextKey = values.splice(0, 1)[0]
//       console.log(nextKey)
//       propsData[`${previousKey}`][`${nextKey}`] = []
//     }


//     let keyDict

//     for (let i of iterator) {
//       keyDict = {}
//       values.map(prop => {
//         if (Array.isArray(prop)) {
//           console.log(prop)
//           console.log('#### calling handleNested from nested ####')
//           // this would only work for one level nested
//           // TODO maybe refactor for infinite recursion
//           console.log(previousKey)
//           console.log(nextKey)
//           console.log(propsData)
//           handleNested(prop, propsData, previousKey, nextKey)
//         } else {
//           console.log(prop)
//           keyDict[`${prop}`] = `sample ${prop}`
//         }
//       })
//       if (firstLevel === true) {
//         propsData[`${previousKey}`].push(keyDict)
//         firstRun = false
//       } else {
//         console.log(propsData)
//         console.log(previousKey)
//         console.log(nextKey)
//         console.log(propsData.experiences)
//         console.log(propsData.experiences.jobs)
//         propsData[`${previousKey}`][`${nextKey}`].push(keyDict)
//       }
//     }
//     return {
//       propsData
//     }
//   }

//   const isNested = props.some(p => Array.isArray(p))
//   const previousKey = null
//   const nextKey = null

//   if (isNested) {
//     console.log('#### is nested ####')
//     handleNested(props, propsData, previousKey, nextKey)
//   } else  {
//     handleArray(props)
//   }
// }
