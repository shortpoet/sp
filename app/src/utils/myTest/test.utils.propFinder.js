import {hardResume} from '@/assets/resume.js'

export const propFinder = (prop) => {
  return {
    propsData: {
      [`${prop}`]: hardResume[`${prop}`]
    }
  }
}

module.exports = propFinder