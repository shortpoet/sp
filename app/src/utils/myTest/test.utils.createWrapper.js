import { createLocalVue } from '@vue/test-utils'
import factory from './test.utils.factory'
import factoryShallow from './test.utils.factoryShallow'

// import VueRouter from 'vue-router'
import Vuex from 'vuex'
import PortalVue from 'portal-vue'

import createStore from '@/store/createStore'
// import createMockRouterConfig from '@/router/createMockRouterConfig'
// import paths from '@/router/paths.js'

import { log } from '@/utils/colorLog'
export const createWrapper  = (
  component,
  /* istanbul ignore next */
  options = {},
  // but why don't i have to ignore the empty  or even unprovided resumeStoreOtions
  // might need to figure this out but setting params to ignore for now
  resumeStoreOptions = {},
  // routerOptions = {},
  isShallow = false,
  ) => {
    const localVue = createLocalVue()
    localVue.use(Vuex)
    localVue.use(PortalVue)
    const mockStoreResume = createStore.createMocks().createStoreResumeMocks(resumeStoreOptions)
    const storeConfig = createStore.createMocks({modules: {resume: mockStoreResume}})

    // Using router is untested makes tests fail
    // and just doesn't work for now
    // localVue.use(VueRouter)
    // const routerConfig = createMockRouterConfig.createMockRouterConfig(routerOptions)
    // const router = new VueRouter(routerConfig)

    // log('green', mockStoreResume)
    // log('yellow', storeConfig.modules.resume)
    const store = new Vuex.Store(storeConfig)
    // log('red', store)
    // log('blue', store._mutations['resume/SET_RESUME_RAW'][0]())
    // console.log(options)
    return isShallow ? 
    factoryShallow(component, {
      localVue,
      store,
      // router,
      ...options
    }) : 
    factory(component, {
      localVue,
      store,
      // router,
      ...options
    })
}

// export default createWrapper
// when this was set to export default a console log(createWrapper) that was in
// test.utils.textMatcher line 19 was being executed from every module that called createWrapper
module.exports = createWrapper