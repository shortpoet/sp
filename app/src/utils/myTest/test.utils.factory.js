import { mount } from '@vue/test-utils'
import { log } from '@/utils/colorLog'
export const factory = (
  component, 
  /* istanbul ignore next */
  options = {
    propsData: {},
    localVue: {},
    mocks: {},
    store: {},
    // context: {}, // only for functional components
    // this only breaks when testing
    // probably should comment out of real func (which this is (now) hehe)
    router: {},
    computed: {},
    stubs: {},
    slots: {},
    scopedSlots: {},
    attrs: {},
    attachToDocument: false,
    attachTo: null, //HTML Element or String
    listeners: {},
    parentComponent: {},
    provide: {}
  }) => {
    // console.log(options)
    // console.log(options.mocks)
    // console.log(options.store)
    // console.log(options.localVue)
    // log('blue', options.stubs)
    // log('green', options.propsData)
    // console.log(mount)
    return mount(component, options)
  }

module.exports = factory

// export default factory