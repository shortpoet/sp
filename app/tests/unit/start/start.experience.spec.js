import * as myTest from '@/utils/myTest'
const { propsMocker, createWrapper } = myTest.default
import StartExperience from '@/components/Resume/Start/StartExperience'
import {cloneDeep} from 'lodash'

describe('StartExperience.vue', () => {

  const component = StartExperience
  // const props = Object.keys(component.props)

  // this was for the _propsMocker attempt
  // const props = [
  //     'experiences',
  //     'type',
  //     ['jobs', 'position', 'company', 'description', '1'],
  //     '4'
  //   ]

  const experiences = ['software', 'language', 'sales', 'hospitality']

  const jobs =  [propsMocker(['position', 'company', 'description', 'startDate', 'endDate']).propsData]
  const propsObject = {
    propsData: {
      experiences: experiences.map(e => {
        return {
          type: e,
          jobs: jobs
        }
      })
    }
  }
  let wrapper = createWrapper(component, propsObject)

  let prop
  let selector
  let mockProp = true

  beforeEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders experience .start-experience-header', () => {

    prop = 'Experience'
    selector = '.start-experience-header'
    expect(wrapper.find(`${selector}`).text()).toMatch(prop)

  })
  it('renders experience .start-experience-type that matches type prop', () => {

    selector = '.start-experience-type'
    const typeArray = wrapper.findAll(`${selector}`)
    for (let i in experiences) {
      expect(typeArray.at(i).text()).toMatch(experiences[i])
    }
  })
  it('renders experience h3 that matches position prop', () => {

    prop = 'sample position'
    selector = 'h3'
    expect(wrapper.find(`${selector}`).text()).toMatch(prop)

  })
  it('renders experience .start-company that matches company prop', () => {

    prop = 'sample company'
    selector = '.start-company'
    expect(wrapper.find(`${selector}`).text()).toMatch(prop)

  })
  it('renders experience span that matches start/end date prop', () => {

    prop = 'sample startDate - sample endDate'
    selector = 'span'
    expect(wrapper.find(`${selector}`).text()).toMatch(prop)

  })
  it('renders experience p that matches text prop accounting for splits', () => {

    let _propsObject = cloneDeep(propsObject)
    _propsObject.propsData.experiences.map(e => {
      e.jobs.map(j => {
        j.description = `description part 1\\n\\rdescription part 2\\n\\rdescription part 3`
      })
    })    
    
    selector = 'p'
    wrapper = createWrapper(component, _propsObject)

    const typeArray = wrapper.findAll(`${selector}`)

    let pIndex = 0
    _propsObject.propsData.experiences.forEach((e, ie) => {
      e.jobs.forEach((j, ij) => {
          for (let id in [...Array(3).keys()]) {
            id = parseInt(id)
            const hashedClass = `job-${ij + 1}-description-${id + 1}`
            expect(typeArray.at(pIndex).attributes().class).toMatch(hashedClass)
            pIndex++
          }
        })
    })

  })
  it('matches experience snapshot', () => {
    
    // same wrapper object from previous tests adds the split descriptions
    // bec a new wrapper was not instantiated
    expect(wrapper.html()).toMatchSnapshot()

  })
})
