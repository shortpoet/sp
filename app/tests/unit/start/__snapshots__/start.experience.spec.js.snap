// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartExperience.vue matches experience snapshot 1`] = `
<section id="experience" class="resume-section p-3 px-lg-5 d-flex flex-column">
  <div class="my-auto">
    <h2 class="heading heading-2 start-experience-header mb-4 mb-md-3">Experience</h2>
    <div class="resume-item d-flex flex-column mb-2">
      <h2 class="heading heading-2 start-experience-type mb-3">software</h2>
      <div class="resume-item d-flex flex-column flex-md-row mb-3">
        <div class="resume-content mr-auto">
          <h3 class="heading heading-3 mb-0">sample position</h3>
          <div class="start-company subheading mb-3">sample company</div>
          <div>
            <p class="job-1-description-1">description part 1</p>
          </div>
          <div>
            <p class="job-1-description-2">description part 2</p>
          </div>
          <div>
            <p class="job-1-description-3">description part 3</p>
          </div>
        </div>
        <div class="resume-date text-md-right mt-1"><span class="text-primary">sample startDate - sample endDate</span></div>
      </div>
    </div>
    <div class="resume-item d-flex flex-column mb-2">
      <h2 class="heading heading-2 start-experience-type mb-3">language</h2>
      <div class="resume-item d-flex flex-column flex-md-row mb-3">
        <div class="resume-content mr-auto">
          <h3 class="heading heading-3 mb-0">sample position</h3>
          <div class="start-company subheading mb-3">sample company</div>
          <div>
            <p class="job-1-description-1">description part 1</p>
          </div>
          <div>
            <p class="job-1-description-2">description part 2</p>
          </div>
          <div>
            <p class="job-1-description-3">description part 3</p>
          </div>
        </div>
        <div class="resume-date text-md-right mt-1"><span class="text-primary">sample startDate - sample endDate</span></div>
      </div>
    </div>
    <div class="resume-item d-flex flex-column mb-2">
      <h2 class="heading heading-2 start-experience-type mb-3">sales</h2>
      <div class="resume-item d-flex flex-column flex-md-row mb-3">
        <div class="resume-content mr-auto">
          <h3 class="heading heading-3 mb-0">sample position</h3>
          <div class="start-company subheading mb-3">sample company</div>
          <div>
            <p class="job-1-description-1">description part 1</p>
          </div>
          <div>
            <p class="job-1-description-2">description part 2</p>
          </div>
          <div>
            <p class="job-1-description-3">description part 3</p>
          </div>
        </div>
        <div class="resume-date text-md-right mt-1"><span class="text-primary">sample startDate - sample endDate</span></div>
      </div>
    </div>
    <div class="resume-item d-flex flex-column mb-2">
      <h2 class="heading heading-2 start-experience-type mb-3">hospitality</h2>
      <div class="resume-item d-flex flex-column flex-md-row mb-3">
        <div class="resume-content mr-auto">
          <h3 class="heading heading-3 mb-0">sample position</h3>
          <div class="start-company subheading mb-3">sample company</div>
          <div>
            <p class="job-1-description-1">description part 1</p>
          </div>
          <div>
            <p class="job-1-description-2">description part 2</p>
          </div>
          <div>
            <p class="job-1-description-3">description part 3</p>
          </div>
        </div>
        <div class="resume-date text-md-right mt-1"><span class="text-primary">sample startDate - sample endDate</span></div>
      </div>
    </div>
  </div>
</section>
`;
